packages/zoe/src/zoe.js

```diff
@@ -275,55 +277,42 @@ const makeZoe = (additionalEndowments = {}) => {
      * @param  {object} installationHandle - the unique handle for the
      * installation
      * @param  {object} terms - arguments to the contract. These
-     * arguments depend on the contract, apart from the `assays`
+     * arguments depend on the contract, apart from the `issuers`
      * property, which is required.
      */
-    makeInstance: (installationHandle, terms) => {
+    makeInstance: (installationHandle, userProvidedTerms) => {
       const { installation } = installationTable.get(installationHandle);
       const instanceHandle = harden({});
       const contractFacet = makeContractFacet(instanceHandle);
 
-      const makeContractInstance = assayRecords => {
-        const synchronousTerms = {
-          ...terms,
-          assays: assayRecords.map(record => record.assay),
+      const makeContractInstance = issuerRecords => {
+        const terms = {
+          ...userProvidedTerms,
+          issuers: issuerRecords.map(record => record.issuer),
         };
-        return installation.makeContract(contractFacet, synchronousTerms);
-      };
 
-      const storeContractInstance = ({
-        invite,
-        publicAPI,
-        terms: contractTerms,
-      }) => {
-        return assayTable
-          .getPromiseForAssayRecords(contractTerms.assays)
-          .then(assayRecords => {
-            const finalAssays = assayRecords.map(record => record.assay);
-
-            const finalTerms = {
-              ...terms,
-              assays: finalAssays,
-            };
-
-            const instanceRecord = harden({
-              installationHandle,
-              publicAPI,
-              assays: finalAssays,
-              terms: finalTerms,
-            });
-
-            instanceTable.create(instanceRecord, instanceHandle);
-            return invite;
-          });
+        const instanceRecord = harden({
+          installationHandle,
+          publicAPI: undefined,
+          issuers: terms.issuers,
+          terms,
+        });
+
+        instanceTable.create(instanceRecord, instanceHandle);
```

Comment:
Doesn't this create a race condition, by registering the instanceRecord in the instanceTable before it is ready, and then updating it later, in a future turn, to a valid state. What happens if it is accessed before you update it? Is there a reason you need to register it before it is ready? I think the old code got that right, by delaying the call to `storeContractInstance` and having it further delay the initial registration until it is fully ready. It was never registered in an invalid state. Did we lose this when the separate `storeContractInstance` function was removed?

Reply:
The only thing that is left out of the first table entry is `publicAPI` - the issuers are already present, so I think we are able to bypass the problems we saw earlier. The reason why we need the instance record there is in case the contract needs to get the issuers from Zoe or the amountMath from Zoe as it is being created, which I think we need for autoswap, but I would have to double check. In any case, if the contract is not able to ask these things when an instance is created, it would be unexpected behavior.

Reply:
Gotcha. It took me awhile, but yes, I think this code is already doing the right thing.

Please put in a comment explaining why you create the instanceTable entry "early", and why it's ok.

Reply:
Added comments